# Melbourne House Price Dataset, predict House Prices based on various features.

Dataset:
https://www.kaggle.com/anthonypino/melbourne-housing-market

In [111]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

import seaborn as sns
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Lasso, Ridge

In [112]:
df = pd.read_csv('Melbourne_housing_FULL.csv')
df

Suburb             Address  Rooms Type      Price Method  \
0      Abbotsford       68 Studley St      2    h        NaN     SS   
1      Abbotsford        85 Turner St      2    h  1480000.0      S   
2      Abbotsford     25 Bloomburg St      2    h  1035000.0      S   
3      Abbotsford  18/659 Victoria St      3    u        NaN     VB   
4      Abbotsford        5 Charles St      3    h  1465000.0     SP   
...           ...                 ...    ...  ...        ...    ...   
34852  Yarraville         13 Burns St      4    h  1480000.0     PI   
34853  Yarraville       29A Murray St      2    h   888000.0     SP   
34854  Yarraville      147A Severn St      2    t   705000.0      S   
34855  Yarraville    12/37 Stephen St      3    h  1140000.0     SP   
34856  Yarraville    3 Tarrengower St      2    h  1020000.0     PI   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis   3/09/2016       2.5    3067.0  ...       1.0  1.0   
1             Biggin   3/12/2016       2.5    3067.0  ...       1.0  1.0   
2             Biggin   4/02/2016       2.5    3067.0  ...       1.0  0.0   
3             Rounds   4/02/2016       2.5    3067.0  ...       2.0  1.0   
4             Biggin   4/03/2017       2.5    3067.0  ...       2.0  0.0   
...              ...         ...       ...       ...  ...       ...  ...   
34852            Jas  24/02/2018       6.3    3013.0  ...       1.0  3.0   
34853        Sweeney  24/02/2018       6.3    3013.0  ...       2.0  1.0   
34854            Jas  24/02/2018       6.3    3013.0  ...       1.0  2.0   
34855  hockingstuart  24/02/2018       6.3    3013.0  ...       NaN  NaN   
34856             RW  24/02/2018       6.3    3013.0  ...       1.0  0.0   

       Landsize  BuildingArea  YearBuilt               CouncilArea Lattitude  \
0         126.0           NaN        NaN        Yarra City Council -37.80140   
1         202.0           NaN        NaN        Yarra City Council -37.79960   
2         156.0          79.0     1900.0        Yarra City Council -37.80790   
3           0.0           NaN        NaN        Yarra City Council -37.81140   
4         134.0         150.0     1900.0        Yarra City Council -37.80930   
...         ...           ...        ...                       ...       ...   
34852     593.0           NaN        NaN  Maribyrnong City Council -37.81053   
34853      98.0         104.0     2018.0  Maribyrnong City Council -37.81551   
34854     220.0         120.0     2000.0  Maribyrnong City Council -37.82286   
34855       NaN           NaN        NaN  Maribyrnong City Council       NaN   
34856     250.0         103.0     1930.0  Maribyrnong City Council -37.81810   

       Longtitude             Regionname Propertycount  
0       144.99580  Northern Metropolitan        4019.0  
1       144.99840  Northern Metropolitan        4019.0  
2       144.99340  Northern Metropolitan        4019.0  
3       145.01160  Northern Metropolitan        4019.0  
4       144.99440  Northern Metropolitan        4019.0  
...           ...                    ...           ...  
34852   144.88467   Western Metropolitan        6543.0  
34853   144.88826   Western Metropolitan        6543.0  
34854   144.87856   Western Metropolitan        6543.0  
34855         NaN   Western Metropolitan        6543.0  
34856   144.89351   Western Metropolitan        6543.0  

[34857 rows x 21 columns]

## Data Analysis & Preprocessing

In [113]:
df.nunique()

Suburb             351
Address          34009
Rooms               12
Type                 3
Price             2871
Method               9
SellerG            388
Date                78
Distance           215
Postcode           211
Bedroom2            15
Bathroom            11
Car                 15
Landsize          1684
BuildingArea       740
YearBuilt          160
CouncilArea         33
Lattitude        13402
Longtitude       14524
Regionname           8
Propertycount      342
dtype: int64

In [114]:
df.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

dropping unnecessary cols 

In [115]:
df.drop(['Address', 'Date', 'Postcode', 'YearBuilt', 'Lattitude', 'Longtitude',], axis='columns', inplace=True)
df

Suburb  Rooms Type      Price Method        SellerG  Distance  \
0      Abbotsford      2    h        NaN     SS         Jellis       2.5   
1      Abbotsford      2    h  1480000.0      S         Biggin       2.5   
2      Abbotsford      2    h  1035000.0      S         Biggin       2.5   
3      Abbotsford      3    u        NaN     VB         Rounds       2.5   
4      Abbotsford      3    h  1465000.0     SP         Biggin       2.5   
...           ...    ...  ...        ...    ...            ...       ...   
34852  Yarraville      4    h  1480000.0     PI            Jas       6.3   
34853  Yarraville      2    h   888000.0     SP        Sweeney       6.3   
34854  Yarraville      2    t   705000.0      S            Jas       6.3   
34855  Yarraville      3    h  1140000.0     SP  hockingstuart       6.3   
34856  Yarraville      2    h  1020000.0     PI             RW       6.3   

       Bedroom2  Bathroom  Car  Landsize  BuildingArea  \
0           2.0       1.0  1.0     126.0           NaN   
1           2.0       1.0  1.0     202.0           NaN   
2           2.0       1.0  0.0     156.0          79.0   
3           3.0       2.0  1.0       0.0           NaN   
4           3.0       2.0  0.0     134.0         150.0   
...         ...       ...  ...       ...           ...   
34852       4.0       1.0  3.0     593.0           NaN   
34853       2.0       2.0  1.0      98.0         104.0   
34854       2.0       1.0  2.0     220.0         120.0   
34855       NaN       NaN  NaN       NaN           NaN   
34856       2.0       1.0  0.0     250.0         103.0   

                    CouncilArea             Regionname  Propertycount  
0            Yarra City Council  Northern Metropolitan         4019.0  
1            Yarra City Council  Northern Metropolitan         4019.0  
2            Yarra City Council  Northern Metropolitan         4019.0  
3            Yarra City Council  Northern Metropolitan         4019.0  
4            Yarra City Council  Northern Metropolitan         4019.0  
...                         ...                    ...            ...  
34852  Maribyrnong City Council   Western Metropolitan         6543.0  
34853  Maribyrnong City Council   Western Metropolitan         6543.0  
34854  Maribyrnong City Council   Western Metropolitan         6543.0  
34855  Maribyrnong City Council   Western Metropolitan         6543.0  
34856  Maribyrnong City Council   Western Metropolitan         6543.0  

[34857 rows x 15 columns]

handling 'NaN' values

In [116]:
df.isnull().sum()

Suburb               0
Rooms                0
Type                 0
Price             7610
Method               0
SellerG              0
Distance             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
CouncilArea          3
Regionname           3
Propertycount        3
dtype: int64

In [117]:
cols_to_fill_zero = ['Distance', 'Bedroom2', 'Bathroom', 'Car', 'Propertycount']
df[cols_to_fill_zero] = df[cols_to_fill_zero].fillna(0)

df.isna().sum()

Suburb               0
Rooms                0
Type                 0
Price             7610
Method               0
SellerG              0
Distance             0
Bedroom2             0
Bathroom             0
Car                  0
Landsize         11810
BuildingArea     21115
CouncilArea          3
Regionname           3
Propertycount        0
dtype: int64

In [118]:
df['Landsize'] = df['Landsize'].fillna(df.Landsize.mean())
df['BuildingArea'] = df['BuildingArea'].fillna(df.BuildingArea.mean())

# df['Price'] = df['Price'].fillna(df.Price.median())

df.isnull().sum()

Suburb              0
Rooms               0
Type                0
Price            7610
Method              0
SellerG             0
Distance            0
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
CouncilArea         3
Regionname          3
Propertycount       0
dtype: int64

In [119]:
df.dropna(inplace=True)

df.isna().sum()

Suburb           0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Distance         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
CouncilArea      0
Regionname       0
Propertycount    0
dtype: int64

In [120]:
df

Suburb  Rooms Type      Price Method        SellerG  Distance  \
1      Abbotsford      2    h  1480000.0      S         Biggin       2.5   
2      Abbotsford      2    h  1035000.0      S         Biggin       2.5   
4      Abbotsford      3    h  1465000.0     SP         Biggin       2.5   
5      Abbotsford      3    h   850000.0     PI         Biggin       2.5   
6      Abbotsford      4    h  1600000.0     VB         Nelson       2.5   
...           ...    ...  ...        ...    ...            ...       ...   
34852  Yarraville      4    h  1480000.0     PI            Jas       6.3   
34853  Yarraville      2    h   888000.0     SP        Sweeney       6.3   
34854  Yarraville      2    t   705000.0      S            Jas       6.3   
34855  Yarraville      3    h  1140000.0     SP  hockingstuart       6.3   
34856  Yarraville      2    h  1020000.0     PI             RW       6.3   

       Bedroom2  Bathroom  Car    Landsize  BuildingArea  \
1           2.0       1.0  1.0  202.000000      160.2564   
2           2.0       1.0  0.0  156.000000       79.0000   
4           3.0       2.0  0.0  134.000000      150.0000   
5           3.0       2.0  1.0   94.000000      160.2564   
6           3.0       1.0  2.0  120.000000      142.0000   
...         ...       ...  ...         ...           ...   
34852       4.0       1.0  3.0  593.000000      160.2564   
34853       2.0       2.0  1.0   98.000000      104.0000   
34854       2.0       1.0  2.0  220.000000      120.0000   
34855       0.0       0.0  0.0  593.598993      160.2564   
34856       2.0       1.0  0.0  250.000000      103.0000   

                    CouncilArea             Regionname  Propertycount  
1            Yarra City Council  Northern Metropolitan         4019.0  
2            Yarra City Council  Northern Metropolitan         4019.0  
4            Yarra City Council  Northern Metropolitan         4019.0  
5            Yarra City Council  Northern Metropolitan         4019.0  
6            Yarra City Council  Northern Metropolitan         4019.0  
...                         ...                    ...            ...  
34852  Maribyrnong City Council   Western Metropolitan         6543.0  
34853  Maribyrnong City Council   Western Metropolitan         6543.0  
34854  Maribyrnong City Council   Western Metropolitan         6543.0  
34855  Maribyrnong City Council   Western Metropolitan         6543.0  
34856  Maribyrnong City Council   Western Metropolitan         6543.0  

[27244 rows x 15 columns]

OneHotEncoding / DummyVars for the categorical features

In [121]:
df = pd.get_dummies(df, drop_first=True)
df = df.astype(float)
df

Rooms      Price  Distance  Bedroom2  Bathroom  Car    Landsize  \
1        2.0  1480000.0       2.5       2.0       1.0  1.0  202.000000   
2        2.0  1035000.0       2.5       2.0       1.0  0.0  156.000000   
4        3.0  1465000.0       2.5       3.0       2.0  0.0  134.000000   
5        3.0   850000.0       2.5       3.0       2.0  1.0   94.000000   
6        4.0  1600000.0       2.5       3.0       1.0  2.0  120.000000   
...      ...        ...       ...       ...       ...  ...         ...   
34852    4.0  1480000.0       6.3       4.0       1.0  3.0  593.000000   
34853    2.0   888000.0       6.3       2.0       2.0  1.0   98.000000   
34854    2.0   705000.0       6.3       2.0       1.0  2.0  220.000000   
34855    3.0  1140000.0       6.3       0.0       0.0  0.0  593.598993   
34856    2.0  1020000.0       6.3       2.0       1.0  0.0  250.000000   

       BuildingArea  Propertycount  Suburb_Aberfeldie  ...  \
1          160.2564         4019.0                0.0  ...   
2           79.0000         4019.0                0.0  ...   
4          150.0000         4019.0                0.0  ...   
5          160.2564         4019.0                0.0  ...   
6          142.0000         4019.0                0.0  ...   
...             ...            ...                ...  ...   
34852      160.2564         6543.0                0.0  ...   
34853      104.0000         6543.0                0.0  ...   
34854      120.0000         6543.0                0.0  ...   
34855      160.2564         6543.0                0.0  ...   
34856      103.0000         6543.0                0.0  ...   

       CouncilArea_Wyndham City Council  CouncilArea_Yarra City Council  \
1                                   0.0                             1.0   
2                                   0.0                             1.0   
4                                   0.0                             1.0   
5                                   0.0                             1.0   
6                                   0.0                             1.0   
...                                 ...                             ...   
34852                               0.0                             0.0   
34853                               0.0                             0.0   
34854                               0.0                             0.0   
34855                               0.0                             0.0   
34856                               0.0                             0.0   

       CouncilArea_Yarra Ranges Shire Council  Regionname_Eastern Victoria  \
1                                         0.0                          0.0   
2                                         0.0                          0.0   
4                                         0.0                          0.0   
5                                         0.0                          0.0   
6                                         0.0                          0.0   
...                                       ...                          ...   
34852                                     0.0                          0.0   
34853                                     0.0                          0.0   
34854                                     0.0                          0.0   
34855                                     0.0                          0.0   
34856                                     0.0                          0.0   

       Regionname_Northern Metropolitan  Regionname_Northern Victoria  \
1                                   1.0                           0.0   
2                                   1.0                           0.0   
4                                   1.0                           0.0   
5                                   1.0                           0.0   
6                                   1.0                           0.0   
...                                 ...                           ...   
34852                   

## Model

In [122]:
X = df.drop('Price', axis='columns')
y = df['Price']

In [123]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [124]:
linear_reg = LinearRegression()
linear_reg.fit(X_train, y_train)

LinearRegression()

In [125]:
linear_reg.score(X_test, y_test)

0.5107699935583883

In [126]:
linear_reg.score(X_train, y_train)

0.6872496733784805

**training score is 68.72% but test score is 51.07% which is very low** --> Normal Regression is clearly over-fitting the data

### Using Lasso (L1 Regularized) Regression Model

In [127]:
lasso_reg = Lasso(alpha=0.1, max_iter=100, tol=0.1)
lasso_reg.fit(X_train, y_train)

c:\Users\Hp\Documents\GitHub\practice-ml\env\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:697: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.407e+15, tolerance: 8.993e+14
  model = cd_fast.enet_coordinate_descent(


Lasso(alpha=0.1, max_iter=100, tol=0.1)

In [128]:
lasso_reg.score(X_test, y_test)

0.5103018631292232

In [129]:
lasso_reg.score(X_train, y_train)

0.6871670904882814

### Using Ridge (L2 Regularized) Regression Model

In [130]:
ridge_reg = Ridge(alpha=1, max_iter=100, tol=0.1)
ridge_reg.fit(X_train, y_train)

Ridge(alpha=1, max_iter=100, tol=0.1)

In [131]:
ridge_reg.score(X_test, y_test)

0.5110893034171372

In [132]:
ridge_reg.score(X_train, y_train)

0.6867174923755653